<a href="https://colab.research.google.com/github/faro7ah/Impact-of-Covid-19-in-a-global-economy-Data-Science-Bootcamp-Capstone-Project-SDA-Coding-Dojo/blob/main/DS_Bootcamp_Impact_of_Covid_19(Capstone_Project).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Impact of Covid-19 Pandemic on the Global Economy  (Capstone Project)**

## **Introduction:**
"The Global COVID-19 pandemic has infected millions of the world's population at a shocking speed. Its
emergence has halted and brought down economic activities across the globe due to the imposition on mobility to
curtail the disease's spread. Over some decades, the world has not witnessed such an experience. However, it has a
significant toll on various economies due to shocks emanating from deteriorating human and health conditions
(World Bank, 2020)".

### **Project Overview:**
In this project, I will try to find how Covid-19 impact the global economy of 170 countries all over the world.


I will approach this problem by firstly, cleaning up the messy data, then doing exploratory analysis(EDA), and finding out who is my most affected countries. Finally, I will create a model to aid us in predicting GDP per capita for countries.

## **Analysis:**
**Part I: Data Exploration:**

In this project, we were given one investigative file. Firstly, I will start analyzing to explore and take a look at the data we have. We have to check if it includes messy data, clean or not. Also, if each column has the right data type, Moreover, if the column called Date is saved as Object or String, then we need to convert it to Date Time.

In [ ]:
import warnings
warnings.filterwarnings("ignore")

In [ ]:
 #install chart_studio package 
#  !pip install chart_studio

In [ ]:
# !pip install plotly

In [ ]:
# install datapane package 
# !pip3 install -U datapane

In [ ]:
#install pycountry and pycountry_convert packages
# !pip install pycountry
# !pip install pycountry_convert

In [ ]:
# import libraries
import pycountry_convert as pc
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import cm
import chart_studio
import chart_studio.plotly as py
import plotly.express as px
import requests
import datetime
import altair as alt
import datapane as dp
import plotly.graph_objs as go
from plotly import tools
import plotly
import colorlover as cl
from IPython.display import HTML
!pip install -q kaggle

In [ ]:
# #Connected to datapane t upload plotly charts
# !datapane login --server='https://datapane.com' --token='#key'


In [ ]:
# #connect to Plotly
# username = '' # your username
# api_key = ''#your Key
# chart_studio.tools.set_credentials_file(username=username, api_key=api_key)

In [ ]:
df1 = pd.read_csv("https://raw.githubusercontent.com/faro7ah/Impact-of-Covid-19-in-a-global-economy-Data-Science-Bootcamp-Capstone-Project-SDA-Coding-Dojo/main/data/transformed_data.csv")

In [ ]:
df2= pd.read_csv('https://raw.githubusercontent.com/faro7ah/Impact-of-Covid-19-in-a-global-economy-Data-Science-Bootcamp-Capstone-Project-SDA-Coding-Dojo/main/data/raw_data.csv')

In [ ]:
df1.head()

CODE      COUNTRY        DATE    HDI   TC   TD  STI        POP    GDPCAP
0  AFG  Afghanistan  2019-12-31  0.498  0.0  0.0  0.0  17.477233  7.497754
1  AFG  Afghanistan  2020-01-01  0.498  0.0  0.0  0.0  17.477233  7.497754
2  AFG  Afghanistan  2020-01-02  0.498  0.0  0.0  0.0  17.477233  7.497754
3  AFG  Afghanistan  2020-01-03  0.498  0.0  0.0  0.0  17.477233  7.497754
4  AFG  Afghanistan  2020-01-04  0.498  0.0  0.0  0.0  17.477233  7.497754

In [ ]:
df2.head()

iso_code     location        date  ...  Unnamed: 11  Unnamed: 12  Unnamed: 13
0      AFG  Afghanistan  2019-12-31  ...        #NUM!    17.477233  7.497754494
1      AFG  Afghanistan  2020-01-01  ...        #NUM!    17.477233  7.497754494
2      AFG  Afghanistan  2020-01-02  ...        #NUM!    17.477233  7.497754494
3      AFG  Afghanistan  2020-01-03  ...        #NUM!    17.477233  7.497754494
4      AFG  Afghanistan  2020-01-04  ...        #NUM!    17.477233  7.497754494

[5 rows x 14 columns]

In [ ]:
#fix pop column in df1 dataframe 
population = df2.population
df1["POP"] = population

In [ ]:
df1.head()

CODE      COUNTRY        DATE    HDI   TC   TD  STI       POP    GDPCAP
0  AFG  Afghanistan  2019-12-31  0.498  0.0  0.0  0.0  38928341  7.497754
1  AFG  Afghanistan  2020-01-01  0.498  0.0  0.0  0.0  38928341  7.497754
2  AFG  Afghanistan  2020-01-02  0.498  0.0  0.0  0.0  38928341  7.497754
3  AFG  Afghanistan  2020-01-03  0.498  0.0  0.0  0.0  38928341  7.497754
4  AFG  Afghanistan  2020-01-04  0.498  0.0  0.0  0.0  38928341  7.497754

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50418 entries, 0 to 50417
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CODE     50418 non-null  object 
 1   COUNTRY  50418 non-null  object 
 2   DATE     50418 non-null  object 
 3   HDI      44216 non-null  float64
 4   TC       50418 non-null  float64
 5   TD       50418 non-null  float64
 6   STI      50418 non-null  float64
 7   POP      50418 non-null  int64  
 8   GDPCAP   50418 non-null  float64
dtypes: float64(5), int64(1), object(3)
memory usage: 3.5+ MB


**Part II: Data understanding**


- **Code** - Represents the ISO Codes for each country.
- **Country** - Contains the names of countries whose impact of  COVID-19.
- **Date** - We are considering cases from December 2019 to October 2020.
- **Month** - Contains months from December 2019 to October 2020.
- **POP** - Shows the population of each country around the world. Crucial for our analysis.
- **GDPCAP** - Represents the GDP per Capita offered by a country. GDP is the Gross Domestic Product. It refers to the sum of money that a particular state can use for the welfare of its citizens. GDP per Capita is the division of the GDP among the citizens. Higher the GDP per Capita, more the country is capable of handling pandemic level situations like COVID-19.
- **HDI** - Human Development Index - a composite measure
of education, life expectancy at birth, and per
capita income indicators.

- **TC** - Refers to Total Cases worldwide. Measured daily, this tally gives us the number of people affected by the COVID-19 pandemic.
-  **TD** - Refers to Total Deaths. Measured daily, recorded from December 31 2019 to October 19 2020.
- **STI** - This is a composite measure
based on nine response indicators, including
school closures, workplace closures, and travel
bans, rescaled to a value from 0 to 100 (100 =
strictest). If policies vary at the subnational level,
the index is shown as the strictest sub-regions
response level

In [ ]:
df1.head()

CODE      COUNTRY        DATE    HDI   TC   TD  STI       POP    GDPCAP
0  AFG  Afghanistan  2019-12-31  0.498  0.0  0.0  0.0  38928341  7.497754
1  AFG  Afghanistan  2020-01-01  0.498  0.0  0.0  0.0  38928341  7.497754
2  AFG  Afghanistan  2020-01-02  0.498  0.0  0.0  0.0  38928341  7.497754
3  AFG  Afghanistan  2020-01-03  0.498  0.0  0.0  0.0  38928341  7.497754
4  AFG  Afghanistan  2020-01-04  0.498  0.0  0.0  0.0  38928341  7.497754

In [ ]:
df1.CODE.unique()

array(['AFG', 'ALB', 'DZA', 'AND', 'AGO', 'AIA', 'ATG', 'ARG', 'ARM',
       'ABW', 'AUS', 'AUT', 'AZE', 'BHS', 'BHR', 'BGD', 'BRB', 'BLR',
       'BEL', 'BLZ', 'BEN', 'BMU', 'BTN', 'BOL', 'BES', 'BIH', 'BWA',
       'BRA', 'VGB', 'BRN', 'BGR', 'BFA', 'BDI', 'KHM', 'CMR', 'CAN',
       'CPV', 'CYM', 'CAF', 'TCD', 'CHL', 'CHN', 'COL', 'COM', 'COG',
       'CRI', 'CIV', 'HRV', 'CUB', 'CUW', 'CYP', 'CZE', 'COD', 'DNK',
       'DJI', 'DMA', 'DOM', 'ECU', 'EGY', 'SLV', 'GNQ', 'ERI', 'EST',
       'ETH', 'FRO', 'FLK', 'FJI', 'FIN', 'FRA', 'PYF', 'GAB', 'GMB',
       'GEO', 'DEU', 'GHA', 'GIB', 'GRC', 'GRL', 'GRD', 'GUM', 'GTM',
       'GGY', 'GIN', 'GNB', 'GUY', 'HTI', 'HND', 'HKG', 'HUN', 'ISL',
       'IND', 'IDN', 'IRN', 'IRQ', 'IRL', 'IMN', 'ISR', 'ITA', 'JAM',
       'JPN', 'JEY', 'JOR', 'KAZ', 'KEN', 'OWID_KOS', 'KWT', 'KGZ', 'LAO',
       'LVA', 'LBN', 'LSO', 'LBR', 'LBY', 'LIE', 'LTU', 'LUX', 'MKD',
       'MDG', 'MWI', 'MYS', 'MDV', 'MLI', 'MLT', 'MRT', 'MUS', 'MEX',
       'MDA', '

In [ ]:
#take a random look of our dataset
df1.sample(10)

CODE        COUNTRY        DATE  ...       STI        POP     GDPCAP
29589  MLT          Malta  2020-07-19  ...  0.000000     441539  10.505432
6778   BRA         Brazil  2020-07-25  ...  4.394696  212559409   9.554175
40086  SEN        Senegal  2020-08-04  ...  3.478467   16743930   7.812208
48173  USA  United States  2020-02-21  ...  1.715598  331002647  10.900906
26512  LBN        Lebanon  2020-08-23  ...  4.171460    6825442   9.500587
38902  LCA    Saint Lucia  2020-07-22  ...  0.000000     183629   9.468993
32417  NAM        Namibia  2020-08-24  ...  4.097506    2540916   9.163438
41055  SGP      Singapore  2020-08-21  ...  3.948355    5850343  11.356685
41920  SOM        Somalia  2020-07-14  ...  3.449352   15893219   0.000000
927    AND        Andorra  2020-06-30  ...  3.660737      77265   0.000000

[10 rows x 9 columns]

In [ ]:
#cheack the number of rows and columns in dataset
df1.shape

(50418, 9)

**Part III: Data Wrangling**

In this part, I will take a deep look through the data and solve some data wrangling to fix issues in columns such as:

- Split date into year, month and day.
- Rename columns into obvious names and change the data type of column date.
- Convert country_code and cuntry columns from object data type to string.
- Create new column for continent

In [ ]:
#function to create new columns contan name of continent
def convert(row):
    # assign the name of the continent for each country if it exists in alpha2
  try:
    country_code = pc.country_name_to_country_alpha2(row.COUNTRY, cn_name_format='default')
    #and Unknown if it is not exist
  except:
    country_code = 'Unknown'
  try:
    conti_code= pc.country_alpha2_to_continent_code(country_code)
  except:
    conti_code= 'Unknown'
    #return name of continent
  return conti_code

In [ ]:
# apply function on our dataset to get name of continent
df1['continent'] = df1.apply(convert, axis=1)

In [ ]:
#check the unique values of continent
df1.continent.unique()

array(['AS', 'EU', 'AF', 'NA', 'SA', 'OC', 'Unknown'], dtype=object)

In this section of the report, we will check for cleanliness,we will trim and clean our dataset for analysis.

In [ ]:
#check for null values
df1.isna().sum()

CODE            0
COUNTRY         0
DATE            0
HDI          6202
TC              0
TD              0
STI             0
POP             0
GDPCAP          0
continent       0
dtype: int64

In [ ]:
#check for duplicated values
df1.duplicated().sum()

0


In this part, I will fix issues in columns such as:
- Convert data type of `DATE` column from object to datetime.
- Split `date` into `year`, `month` and `day`.
- Rename columns into obvious names.
- Convert `country_code` and `cuntry` columns from object data type to string.


In [ ]:
#convert ds to datatime
df1['DATE'] = pd.to_datetime(df1['DATE'])

In [ ]:
#split year into new column
df1['year'] = df1['DATE'].dt.year
#split month into new column
df1['month'] = df1['DATE'].dt.month
#split day into new column
df1['day'] = df1['DATE'].dt.day

In [ ]:
 #rename columns into obvious column name
 df1.rename(columns={'CODE': 'country_code', 'COUNTRY': 'country', 'DATE': 'date',
                    'HDI': 'human_development_index',
                    'TC': 'total_cases', 'TD':'total_deaths',
                    'POP':'population','STI':'stringency_index', 'GDPCAP':'gdp_per_capita'}, inplace=True)

In [ ]:
#check column names
df1.head()

country_code      country       date  ...  year  month  day
0          AFG  Afghanistan 2019-12-31  ...  2019     12   31
1          AFG  Afghanistan 2020-01-01  ...  2020      1    1
2          AFG  Afghanistan 2020-01-02  ...  2020      1    2
3          AFG  Afghanistan 2020-01-03  ...  2020      1    3
4          AFG  Afghanistan 2020-01-04  ...  2020      1    4

[5 rows x 13 columns]

## **EDA(Exploratory Data Analysis):** 
- In this part, we will plot interactive charts to make interacting with our dataset.

In [ ]:
#display column name 
df1.columns

Index(['country_code', 'country', 'date', 'human_development_index',
       'total_cases', 'total_deaths', 'stringency_index', 'population',
       'gdp_per_capita', 'continent', 'year', 'month', 'day'],
      dtype='object')

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50418 entries, 0 to 50417
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   country_code             50418 non-null  object        
 1   country                  50418 non-null  object        
 2   date                     50418 non-null  datetime64[ns]
 3   human_development_index  44216 non-null  float64       
 4   total_cases              50418 non-null  float64       
 5   total_deaths             50418 non-null  float64       
 6   stringency_index         50418 non-null  float64       
 7   population               50418 non-null  int64         
 8   gdp_per_capita           50418 non-null  float64       
 9   continent                50418 non-null  object        
 10  year                     50418 non-null  int64         
 11  month                    50418 non-null  int64         
 12  day                      50418 n

In [ ]:
#for loop to determine numeric columns
numeric_column = [col for col in df1.select_dtypes(include=['float64']).columns]
numeric_column

['human_development_index',
 'total_cases',
 'total_deaths',
 'stringency_index',
 'gdp_per_capita']

In [ ]:
#numeric columns in dataframe
numeric_cols =df1[numeric_column]
numeric_cols.head()

human_development_index  total_cases  ...  stringency_index  gdp_per_capita
0                    0.498          0.0  ...               0.0        7.497754
1                    0.498          0.0  ...               0.0        7.497754
2                    0.498          0.0  ...               0.0        7.497754
3                    0.498          0.0  ...               0.0        7.497754
4                    0.498          0.0  ...               0.0        7.497754

[5 rows x 5 columns]

### Box plot for human_development_index, total_cases, total_deaths, stringency_index, and gdp_per_capita:

In [ ]:
# #interactive Boxplot to check Outliers in numeric column
fig = go.Figure()
#  #for loop for numeric columns
for col in numeric_cols:
#    #boxplot 
    fig.add_trace(go.Box(y=numeric_cols[col][:8000].values, name=numeric_cols[col].name))
    fig.update_layout(
      height=800,plot_bgcolor='rgb(1000,1000,1000)',
      title_text='Box plot for human_development_index, total_cases, total_deaths, stringency_index, and gdp_per_capita')
#disply chart    
fig.show()
py.plot(fig, filename = 'gdp_per_cap', auto_open=True) #Publish the chart


'https://plotly.com/~farah12/40/'

### It is illustrated from the previous chart:
### - **Gdp_per_capita:** the median is around 9.36 and most of GDP_percapita falls between 8 and $10.3 thousand approximatelly, and we can notice one outlier which is 0.

### - **Total_cases:** We can immediately draw a few conclusions from this:
- The total_cases was usually below 16 thousand.
- In 75% of all cases, the total_cases was smaller than 11 thousand  (3rd Quantile)
- 50% of total_cases gave at least 7 thousand (Median).




---

---





In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50418 entries, 0 to 50417
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   country_code             50418 non-null  object        
 1   country                  50418 non-null  object        
 2   date                     50418 non-null  datetime64[ns]
 3   human_development_index  44216 non-null  float64       
 4   total_cases              50418 non-null  float64       
 5   total_deaths             50418 non-null  float64       
 6   stringency_index         50418 non-null  float64       
 7   population               50418 non-null  int64         
 8   gdp_per_capita           50418 non-null  float64       
 9   continent                50418 non-null  object        
 10  year                     50418 non-null  int64         
 11  month                    50418 non-null  int64         
 12  day                      50418 n

### A pairs plot to see both distributions of single global economy feature and relationships between these features that impacted by COVID-19:

In [ ]:
import plotly.express as px
fig = px.scatter_matrix(df1.sample(20))
# chart title and size
fig.update_layout(title="Pair Plot For Impact of Covid-19 Pandemic on the Global Economy ",
                     dragmode='select',
                     width=3000,
                     height=2300,
                     hovermode='closest')
fig.show()
report = dp.Report(dp.Plot(fig) ) #Create a report
report.publish(name='Pair Plot For Impact of Covid-19 Pandemic on the Global Economy', open=True, visibility='PUBLIC') #Publish the report

Publishing report and associated data - please wait..
Report successfully published at https://datapane.com/u/farahh212/reports/pair-plot-for-impact-of-covid-19-pandemic-on-the-global-economy/




---

---





### Bar polt to display the rate of GDP for top 50 countries around the world:

In [ ]:
#groupby countries to find gdp for top 50 countries
gpd_per_couuntry= df1.groupby(['country'])['gdp_per_capita'].sum()
#sort values
gpd_per_couuntry.sort_values()

country
Isle of Man                    0.000000
New Caledonia                  0.000000
Northern Mariana Islands       0.000000
Montserrat                     0.000000
Monaco                         0.000000
                               ...     
United Arab Emirates        3268.344692
Ireland                     3268.527298
Singapore                   3338.865511
Luxembourg                  3367.476814
Qatar                       3430.797317
Name: gdp_per_capita, Length: 210, dtype: float64

In [ ]:
#Datafram for top 50 countries according GDP per capita
gpd_per_couuntry_ =gpd_per_couuntry.sort_values(ascending=False).head(50)
gpd_per_couuntry_ = gpd_per_couuntry_.to_frame(name='gpd_per_couuntry').reset_index()

In [ ]:
#type(gpd_per_couuntry_)

In [ ]:
gpd_per_couuntry_

country  gpd_per_couuntry
0                  Qatar       3430.797317
1             Luxembourg       3367.476814
2              Singapore       3338.865511
3                Ireland       3268.527298
4   United Arab Emirates       3268.344692
5                 Kuwait       3260.539830
6                 Norway       3257.244157
7            Switzerland       3221.645141
8             San Marino       3218.821733
9          United States       3204.866235
10           Netherlands       3171.893332
11               Denmark       3160.830738
12               Iceland       3159.571263
13               Austria       3152.878085
14               Germany       3151.532756
15             Australia       3147.734723
16                Canada       3143.549321
17               Bahrain       3138.653810
18                Sweden       3130.235554
19               Finland       3119.684445
20        United Kingdom       3113.591339
21               Belgium       3113.007214
22                 Japan       3107.983932
23                France       3104.979411
24                  Oman       3100.026237
25           New Zealand       3085.134838
26           South Korea       3083.930910
27                 Italy       3077.995300
28                Israel       3060.029769
29                 Spain       3059.533683
30        Czech Republic       3055.321124
31             Lithuania       3026.132519
32               Estonia       3025.703884
33              Malaysia       2997.759754
34                Russia       2974.464200
35                Greece       2972.181181
36               Romania       2956.691838
37               Croatia       2948.463904
38                  Iran       2897.820776
39                Mexico       2869.606873
40               Belarus       2866.735360
41              Thailand       2851.079574
42            Azerbaijan       2843.204007
43                  Iraq       2839.781124
44                 China       2833.036143
45    Dominican Republic       2819.117719
46                Brazil       2808.927411
47               Algeria       2804.947935
48               Lebanon       2793.172440
49             Macedonia       2787.479613

In [ ]:
#interactive bar plot for top 50 countries according the GDP per capita 
fig = px.bar(gpd_per_couuntry_, y='gpd_per_couuntry', x='country', text='gpd_per_couuntry', color_discrete_sequence=px.colors.qualitative.Plotly)
#round GPD into 2 values 
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')#, marker_color='purple')
#chart title and colors
fig.update_layout(title="GPD Per Capita According Each Coutry", width=1200,height=800, plot_bgcolor='rgb(1000,1000,1000)')
#display chart
fig.show()
#publish chart 
py.plot(fig, filename = 'GDP Per Capita According Each Coutry', auto_open=True)

'https://plotly.com/~farah12/21/'

### From the previous chart, we can see that `Qatar, Luxembourg, Singapore, Ireland, United Arab Emirates, and Kuwait` recorded the top 5 rates of GDP per capita with `3430.797317, 3367.476814, 3338.865511, 3268.527298, 3268.344692, 3260.539830` respectively.



---

---









### The next pie charts illustrate the rate of total deaths and total cases under this global pandemic:

In [ ]:
#gropby countries for total deaths rate
total_deaths_per_country= df1.groupby(['country'])['total_deaths'].sum()
total_deaths_per_country.sort_values()

country
Monaco                       0.000000
Timor                        0.000000
Liechtenstein                0.000000
British Virgin Islands       0.000000
Gibraltar                    0.000000
                             ...     
France                    2203.766710
China                     2205.740166
United Kingdom            2209.519314
Italy                     2327.710204
United States             2480.576064
Name: total_deaths, Length: 210, dtype: float64

In [ ]:
#dataframe for countries and rate of total deaths
total_deaths_per_country_ =total_deaths_per_country.sort_values(ascending=False).head(20)
total_deaths_per_country_ = total_deaths_per_country_.to_frame(name='total_deaths_per_country').reset_index()

In [ ]:
total_deaths_per_country_

country  total_deaths_per_country
0    United States               2480.576064
1            Italy               2327.710204
2   United Kingdom               2209.519314
3            China               2205.740166
4           France               2203.766710
5            Spain               2197.919025
6           Brazil               2166.965888
7             Iran               2112.453855
8           Mexico               1963.633939
9            India               1933.999000
10         Belgium               1913.059152
11         Germany               1882.624124
12     Netherlands               1829.550001
13            Peru               1805.656658
14          Canada               1794.530769
15          Turkey               1744.453495
16          Russia               1742.065306
17          Sweden               1730.047120
18         Ecuador               1692.027530
19       Indonesia               1660.869037

In [ ]:
#group by countries for total cases rate
total_cases_per_country= df1.groupby(['country'])['total_cases'].sum()
total_cases_per_country.sort_values()

country
Hong Kong                             0.000000
Solomon Islands                       3.988984
Anguilla                            224.574488
Falkland Islands                    493.373289
Bonaire Sint Eustatius and Saba     497.356421
                                      ...     
Spain                              2843.820343
Italy                              2873.314489
Brazil                             2913.256895
China                              3128.587021
United States                      3343.050755
Name: total_cases, Length: 210, dtype: float64

In [ ]:
#dataframe for total cases rate on each country 
total_cases_per_country_ =total_cases_per_country.sort_values(ascending=False).head(20)
total_cases_per_country_ = total_cases_per_country_.to_frame(name='total_cases_per_country').reset_index()

In [ ]:
#pie plot for total_cases rate according each country
fig = px.pie(total_cases_per_country_, values='total_cases_per_country', names='country')
fig.update_traces(textposition='inside', marker= dict(line=dict(color='#000000', width=2)))
#chart title
fig.update_layout(title="Rate Of Total Cases According Each county")
#display chart
fig.show()
 #Publish chart
py.plot(fig, filename = 'Rate Of Total Cases According Each county', auto_open=True)

'https://plotly.com/~farah12/23/'



---

---





In [ ]:
#pie chart for total deaths per country
fig = px.pie(total_deaths_per_country_, values='total_deaths_per_country', names='country')
fig.update_traces(textposition='inside', marker= dict(line=dict(color='#000000', width=2)))
#chart title
fig.update_layout(title="Rate Of Deaths According Each county")
#display chart
fig.show()
#Publish chart 
py.plot(fig, filename = 'Rate Of Deaths According Each county', auto_open=True)

'https://plotly.com/~farah12/25/'

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50418 entries, 0 to 50417
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   country_code             50418 non-null  object        
 1   country                  50418 non-null  object        
 2   date                     50418 non-null  datetime64[ns]
 3   human_development_index  44216 non-null  float64       
 4   total_cases              50418 non-null  float64       
 5   total_deaths             50418 non-null  float64       
 6   stringency_index         50418 non-null  float64       
 7   population               50418 non-null  int64         
 8   gdp_per_capita           50418 non-null  float64       
 9   continent                50418 non-null  object        
 10  year                     50418 non-null  int64         
 11  month                    50418 non-null  int64         
 12  day                      50418 n

In [ ]:
#group by countries for total cases rate
gdp_per_human_development_index= df1.groupby(['gdp_per_capita'])['human_development_index'].sum()
gdp_per_human_development_index.sort_values()

gdp_per_capita
10.500323      0.000
10.833076      0.000
10.948373      0.000
10.817837      0.000
10.464379      0.000
              ...   
11.117440    275.772
10.706581    276.066
10.957977    277.536
11.079062    280.182
0.000000     682.351
Name: human_development_index, Length: 185, dtype: float64

In [ ]:
#create new dataframe for scatter plot
#groupby df by country
multi_df = df1.groupby(['human_development_index', 'country'],
                       #calculate sum and avg of gdp
                       as_index=False).agg({'gdp_per_capita': ['sum', 'mean']})
multi_df.columns =[
                   '_'.join(col).rstrip('_') for col in multi_df.columns.values
]

In [ ]:
multi_df

human_development_index  ... gdp_per_capita_mean
0                      0.000  ...            9.189712
1                      0.354  ...            6.830874
2                      0.367  ...            6.494117
3                      0.388  ...            7.358760
4                      0.404  ...            7.477691
..                       ...  ...                 ...
177                    0.936  ...           10.719499
178                    0.938  ...           11.117440
179                    0.939  ...           10.706581
180                    0.944  ...           10.957977
181                    0.953  ...           11.079062

[182 rows x 4 columns]



---

---





### Scatter plot to show relationships among GDP per capita and human_development_index under Covid-19:



In [ ]:
#interactive scatter plot
fig = px.scatter(multi_df, y="gdp_per_capita_sum", x="human_development_index",hover_data=['human_development_index', 'country', 'gdp_per_capita_mean'], color='country',  log_x=True, size_max=60)
#chart title and background color 
fig.update_layout(title="The Relationship Among GDP Per Capitaa And Human Development Index For Each Country Under Covid-19", plot_bgcolor='rgb(1000,1000,1000)')
#display chart
fig.show()
#Publish chart
py.plot(fig, filename = 'The Relationship Among GDP Per Capitaa And Human Development Index', auto_open=True)


'https://plotly.com/~farah12/27/'

### According to what is shown, the countries that increased the human development index recorded the highest rate of GDP per capita such as Norway and Switzerland and the United States. Moreover, in this scatter plot we have a Strong Positive relationship, which is mean; when the human development index increased then GDP per capita will definitely increase.



---

---





In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50418 entries, 0 to 50417
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   country_code             50418 non-null  object        
 1   country                  50418 non-null  object        
 2   date                     50418 non-null  datetime64[ns]
 3   human_development_index  44216 non-null  float64       
 4   total_cases              50418 non-null  float64       
 5   total_deaths             50418 non-null  float64       
 6   stringency_index         50418 non-null  float64       
 7   population               50418 non-null  int64         
 8   gdp_per_capita           50418 non-null  float64       
 9   continent                50418 non-null  object        
 10  year                     50418 non-null  int64         
 11  month                    50418 non-null  int64         
 12  day                      50418 n

In [ ]:
#new DF for 'population','country_code','country','month','stringency_index', and sum of gdp per capita
multi_df1 = df1.groupby(['population','country_code','country','month','stringency_index'],
                       as_index=False).agg({'gdp_per_capita': ['sum']})
multi_df1.columns =[
                   '_'.join(col).rstrip('_') for col in multi_df1.columns.values
]

In [ ]:
multi_df1

population country_code  ... stringency_index  gdp_per_capita_sum
0            809          VAT  ...         0.000000            0.000000
1            809          VAT  ...         0.000000            0.000000
2            809          VAT  ...         0.000000            0.000000
3            809          VAT  ...         0.000000            0.000000
4            809          VAT  ...         0.000000            0.000000
...          ...          ...  ...              ...                 ...
4456  1439323774          CHN  ...         4.359781           86.725596
4457  1439323774          CHN  ...         4.405987          105.997951
4458  1439323774          CHN  ...         0.000000          115.634128
4459  1439323774          CHN  ...         3.992127           67.453241
4460  1439323774          CHN  ...         0.000000            9.636177

[4461 rows x 6 columns]

### World Average GPD Per Capita:

In [ ]:
#avg_gdp = (df1['gdp_per_capita']*df1['population']).sum()/df1['population'].sum()
multi_df1['month'] = multi_df1['month'].replace({1:"Jan 2020", 2:"Feb 2020", 3:"Mar 2020", 4:"Apr 2020",
                                                   5:"May 2020", 6:"Jun 2020", 7:"Jul 2020", 8:"Aug 2020",
                                                   9:"Sep 2020", 10:"Oct 2020", 12:"Dec 2019",})
fig = px.choropleth(multi_df1, locations="country_code", color='gdp_per_capita_sum', hover_data=['country','stringency_index', 'gdp_per_capita_sum'] , animation_frame='month', 
                     color_continuous_scale=px.colors.sequential.dense,
                     projection ='natural earth',
                     title="The Rate Of Stringency Index and GDP per capita According to Each Country")
fig.update_layout(width=800,height=900)
#display chart
fig.show()
#publish chart
py.plot(fig, filename = 'The Rate Of Stringency Index and GDP', auto_open=True)

'https://plotly.com/~farah12/33/'

### In thi previous interactive map chart, we want to discover the average GDP, rate of Total Cases and Stringency Index according to each month in each country s under this global pandemic.

# Prepare data for modeling:

### Remove outlier from GPD:
In this part, we will remove outliers from our target(GDP per capita), because we found that does not make sense to include 0 values on the GDP per capita for countries.

**Baseline Model:**

In [ ]:
# Baseline Score
print("Model Baseline")
baseline = df1["gdp_per_capita"].value_counts(normalize=True)[0]
baseline

Model Baseline


0.11329287159347852

In [ ]:
#removing outliers from thr target column
df1.drop(df1.index[df1['gdp_per_capita'] ==0], inplace = True)

In [ ]:
#check null values
df1.isna().sum()

country_code                  0
country                       0
date                          0
human_development_index    1370
total_cases                   0
total_deaths                  0
stringency_index              0
population                    0
gdp_per_capita                0
continent                     0
year                          0
month                         0
day                           0
dtype: int64

In [ ]:
 #fill NAN values with median
df1['human_development_index'].fillna(df1['human_development_index'].median(), inplace=True)

In [ ]:
#drop useless columns
df1.drop(columns=['country_code', 'date', 'continent'], inplace=True)

In [ ]:
dff2 = df1.copy()

In [ ]:
dff2.drop(columns=['country'], inplace=True)

In [ ]:
#plot the corr for our dataframe, interactive heatmap
data = [
      go.Heatmap(
          #corr column and values
          z= dff2.corr().values,
          x=dff2.columns.values,
          y=dff2.columns.values,
          #heatmap color
          colorscale='BuPu',
          reversescale = False)
  ]
#heatmap title and size
layout = go.Layout(
      title="Features' Correlation Of COVID-19 Economic Impact",
      xaxis = dict(ticks='', nticks=36),
      yaxis = dict(ticks='' ),
      width = 900, height = 700)
#display heatmap
fig = go.Figure(data=data, layout=layout)
fig.show()
py.plot(fig, filename = "Features' Correlation", auto_open=True)


# report = dp.Report(dp.Plot(fig) ) #Create a report
# report.publish(name='Correlation Of COVID-19 Economic Impact Features', open=True, visibility='PRIVATE') #Publish the report

'https://plotly.com/~farah12/35/'

In [ ]:
#onehotencoder for catecorical columns
df1=pd.get_dummies(df1)

In [ ]:
#check shapd of data after get_dummies
df1.shape

(44706, 193)

In [ ]:
df1.head()

human_development_index  total_cases  ...  country_Zambia  country_Zimbabwe
0                    0.498          0.0  ...               0                 0
1                    0.498          0.0  ...               0                 0
2                    0.498          0.0  ...               0                 0
3                    0.498          0.0  ...               0                 0
4                    0.498          0.0  ...               0                 0

[5 rows x 193 columns]

In [ ]:
#split our data into dependent(X) and independent(y) column
X=df1.drop(['gdp_per_capita'],axis=1)
y=df1['gdp_per_capita']

### Feature Selections: 

In [ ]:
from sklearn.linear_model import Lasso
from sklearn.feature_selection import SelectFromModel
#lasso performs feature selection that can pick relevant features that will be useful for modelling
feature_sel_model = SelectFromModel(Lasso(alpha=0.005, random_state=42))

In [ ]:
#fit data into feature selection model
feature_sel_model.fit(X,y)

SelectFromModel(estimator=Lasso(alpha=0.005, copy_X=True, fit_intercept=True,
                                max_iter=1000, normalize=False, positive=False,
                                precompute=False, random_state=42,
                                selection='cyclic', tol=0.0001,
                                warm_start=False),
                max_features=None, norm_order=1, prefit=False, threshold=None)

In [ ]:
#check for best feature
feature_sel_model.get_support()

array([ True,  True,  True,  True, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False,  True,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True,  True,
       False, False, False, False, False, False, False, False,  True,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [ ]:
#best feature colusmns
cols=X.columns

In [ ]:
#check for total features, and the best feature for modeling 
selected_feature = cols[feature_sel_model.get_support()]

In [ ]:
#print total feature and selected feature
print("total_features: {}".format(X.shape[1]))
print("selected features: {}".format(len(selected_feature)))

total_features: 192
selected features: 13


In [ ]:
selected_feature

Index(['human_development_index', 'total_cases', 'total_deaths',
       'stringency_index', 'country_Bermuda', 'country_Cayman Islands',
       'country_Equatorial Guinea', 'country_Kosovo', 'country_Kuwait',
       'country_Luxembourg', 'country_Qatar', 'country_San Marino',
       'country_United Arab Emirates'],
      dtype='object')

In [ ]:
#fit selected feature into X
X=X[selected_feature]

### Split our data into tain_test_split:

In [ ]:
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2, random_state=100)

### Since we have a regression problem, we will use MAE and RMSE to evaluate our models.

In [ ]:
from sklearn import metrics
#function to fit train data into model, apply cross validation, and calculate MAE , MSE and RMSE
def predict(ml_model):
  #fit trainin set into model
  model = ml_model.fit(X_train, y_train)
  #cross validate model
  coross_fal = cross_val_score(model,X_train, y_train, cv=10, n_jobs=-1)

  #predict x_test
  predictions = model.predict(X_test)
  #print("predictions: {}".format(predictions))
  print("\n")
  #caluculate R2  score, MAE, RMSE
  r2_score = metrics.r2_score(y_test, predictions)
  print("R2 Score: {}".format(r2_score))
  print("MAE Score:", metrics.mean_absolute_error(y_test, predictions))
  print("RMSE Score:",np.sqrt(metrics.mean_absolute_error(y_test, predictions)))
  

### 1- LinearRegression Model:

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
LinearRegression()

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
predict(LinearRegression())



R2 Score: 0.9261595229666666
MAE Score: 0.2402406844123298
RMSE Score: 0.49014353450018067


### Grid search to optimized our model:

In [ ]:
model =LinearRegression()
param_grid ={'fit_intercept': [True, False],
             'n_jobs':[-1,None],
             'normalize':[True,False]}
grid = GridSearchCV(estimator = model, param_grid = param_grid, cv = 5, n_jobs = -1)
grid_result = grid.fit(X_train, y_train)


In [ ]:
grid_result.best_params_

{'fit_intercept': True, 'n_jobs': -1, 'normalize': True}

In [ ]:
grid_result.best_score_

0.9283885126876091

### After using grid search we can notice that we have the same result.

### In this part of data modelling, we used LinearRegression and RandomForest. Moreover, to evaluate our model performance, we used MAE because it is conceptually the easiest evaluation metric for regression problem and it answers the question "How far were you off in your predictions, on average?"
### Here we obtained 0.24 which is a good score.
### Finally, we used RMSE, which is the most commonly used in the regression. Moreover, it is defined as the square root of the average squared distance between the actual score and prediction score. Here we have 0.49.
### In the Baseline model, we got 0.11, furthermore, the previous result is better than the Baseline model.

#Improvements: 

### In my point of view, I think I got to a point where we have good results and we understand the data we have very well. But to make our results even better, I would try to improve data collection I will also try to get even more data for the last 5 years to compare the GPD per capita before and after this global pandemic Because collecting more data is always lead us to improve our model results.

# Conclusion:
### To conclude, In this project, we tried to analyze and build models to predict the best GDP per capita. Firstly, I explored our dataset and make some data wrangling before prepare data for modelling Then I did EDA on the data after cleaning. From that analysis, I found out that GDP per capita changing from country to country depends on some features such as human development index, total cases, total deaths, stringency index and rate of population. I dug deep to see who countries affected by Covid-19, and I found that most of the countries all over the world affected by this global pandemic according to the rate of cases and deaths. Finally, My plan for the next step, comparing these outcomes with the global economy for the last 5 years all over the world then we will find significant results under Covid-19.

## Resources:
- [THE IMPACT OF COVID-19 PANDEMIC ON THE GLOBAL ECONOMY.](http://www.conscientiabeam.com/pdf-files/eco/29/EFL20218(1)32-43.pdf)

- [Impact of Covid-19 Pandemic on the Global Economy.](https://www.kaggle.com/shashwatwork/impact-of-covid19-pandemic-on-the-global-economy/)

- [Plotly.](https://chart-studio.plotly.com/)
- [Lasso.](https://medium.com/all-about-ml/search?q=Lasso)
- [LinearRegression.](https://medium.com/all-about-ml/linear-regression-d41a6a5dcab6)